### PR TITLE
Nam Nguyen Code Review

### PR DESCRIPTION
**Pull request by tnam-nguyen**
Code Review Pull Request for Nam Nguyen

Source Files that I have worked on:

1. Sprint 1: 

- gameplay.js, constant.js:  Created a gameplay 
- skinPage.js, settingPage.js: Created a horizontal menu and integrated the settings to the gameplay

2. Sprint 2: 
- .travis.yml, requirements.txt : Initialed the CI service to run both the Frontend and Backend 
- Form.js: Created a Form to get the username
- test_backend.py: Created Backend Unit Tests for The CI Service
- model_mongodb.py: Integrated the sort function, fixed indentation and added comments to pass the PEP8 Style Check
- sample_backend.py: fixed indentation and added comments to pass the PEP8 Style Check

3. Sprint 3:
- App.test.js: Frontend Unit Test

### CODE REVIEW

File: src/Form.js [hunk 1]
```diff
@@ -1,46 +1,51 @@
 import React, {Component} from 'react'
-import { FaPlayCircle, FaListOl, FaHourglassStart } from 'react-icons/fa';
-import { Link } from 'react-router-dom';
+import {FaPlayCircle, FaListOl, FaHourglassStart} from 'react-icons/fa';
+import {Link} from 'react-router-dom';
+import {settings} from "./constants";
 
 class Form extends Component {
```
Comment (eddyiwnl) [inline src/Form.js:6]: Extra blank lines in Form.js could be removed.

File: rest-api/test_backend.py [hunk 2]
```diff
@@ -8,31 +8,7 @@ def test_get_leaderboard():
...
-        "_id": "602d11996fc98b2990e2b683",
-        "name": "Brett",
-        "score": 19044
-      },
-    ]
-
     assert resp.status_code == 200
-    assert expected == resp.get_json()["leaderboard"][:4]
 
 def test_post_delete_leaderboard():
```
Comment (eddyiwnl) [inline rest-api/test_backend.py:39]: There could be more backend tests, maybe some for the GET methods.

File: src/App.test.js [hunk 3]
```diff
@@ -1,9 +1,51 @@
...
+let pathMap = {};
+describe('routes using array of routers', () => {
+  beforeAll(() => {
+    const component = shallow(<App/>);
+    pathMap = component.find(Route).reduce((pathMap, route) => {
+        const routeProps = route.props();
+        pathMap[routeProps.path] = routeProps.component;
+        return pathMap;
+      }, {});
+  })
```
Comment (eddyiwnl) [inline src/App.test.js:30]: Good job with the thorough front-end testing!

File: src/gameplay.js [hunk 4]
```diff
@@ -116,21 +123,27 @@ const Gameplay = () => {
...
+            <div className="gameplay-div">
+                <div role="button" tabIndex="0" className="game-screen" onKeyDown={e => moveSnake(e)}>
+                    <div className="rectangle"/>
+                    <div className="score-text"> SCORE: {score}</div>
+                    <div className="name-text"> Welcome, {settings.playerName}</div>
+                    <canvas
+                        style={{backgroundColor: "lightgreen", position: "absolute", top: "180px"}}
+                        ref={canvasRef}
+                        width={`${CANVAS_SIZE[0]}px`}
+                        height={`${CANVAS_SIZE[1]}px`}
```
Comment (eddyiwnl) [inline src/gameplay.js:139]: I would recommend moving inline CSS elements to app.css

File: src/gameplay.js [hunk 5]
```diff
@@ -83,14 +86,18 @@ const Gameplay = () => {
         return false;
     }
     const checkAppleCollision = newSnake => {
```
Comment (eddyiwnl) [inline src/gameplay.js:88]: Gameplay logic looks good to me!

**Review by eddyiwnl — COMMENTED** [review]
Goodjob with the testing and the code style!

File: src/App.test.js (hunk 3 above)
Comment (tnam-nguyen) [reply src/App.test.js:30]: Thank you!

File: src/gameplay.js (hunk 5 above)
Comment (tnam-nguyen) [reply src/gameplay.js:88]: Thank you!

File: src/gameplay.js (hunk 4 above)
Comment (tnam-nguyen) [reply src/gameplay.js:139]: It looks like to me that the Canvas HTML element ignores the class imported from the external CSS style sheet, so I have to keep the inline CSS style for now

File: src/Form.js (hunk 1 above)
Comment (tnam-nguyen) [reply src/Form.js:6]: Fix those under the last commit, "Update Backend Testing for GET requests. Delete blank lines in Form.js"

File: rest-api/test_backend.py (hunk 2 above)
Comment (tnam-nguyen) [reply rest-api/test_backend.py:39]: Since the content of the leaderboard in the database usually changes, It is quite difficult to test for the GET method and see if GET retrieve the correct data.  For the improved Unit Test for the GET method, I first POST a few high scores to the leaderboard on the database, and test to see if GET correctly retrieve those high scores. After, I DELETE any scores I just added to preserve the old content of the leaderboard. You can check these changes under the last commit, "Update Backend Testing for GET requests. Delete blank lines in Form.js"

File: src/gameplay.js [hunk 6]
```diff
@@ -36,31 +39,31 @@ const Gameplay = () => {
     }
 
 
-    const writeLeaderboard = () =>
-    {
+    const writeLeaderboard = () => {
```
Comment (mdg333) [inline src/gameplay.js:42]: Thank you for adding the score posting functionality.

File: .travis.yml [hunk 7]
```diff
@@ -27,6 +23,8 @@ jobs:
         - npm install react-bootstrap --save
         - npm install react-horizontal-scrolling-menu --save
         - npm install axios --save
+        - npm install enzyme --save
```
Comment (mdg333) [inline .travis.yml:26]: Your work with the Travis CI service & misc. documentation does not go unappreciated, thank you!

File: rest-api/test_backend.py [hunk 8]
```diff
@@ -53,3 +22,49 @@ def test_post_delete_leaderboard():
...
+    # Initial test data to POST
+    data1 = {
+      "name": "Player1",
+      "score": 1000000000
+    }
+    data2 = {
+      "name": "Player2",
+      "score": 34325200
+    }
+    data3 = {
```
Comment (mdg333) [inline rest-api/test_backend.py:73]: If testing multiple posts, I suggest adding special cases to test if the leaderboard can handle erroneous data such as negative scores or invalid characters in the name.

**Review by mdg333 — COMMENTED** [review]
Good work Nam! Thank you for proactively setting the pace for sprint 2 and filling any feature gaps.

File: src/gameplay.js (hunk 6 above)
Comment (tnam-nguyen) [reply src/gameplay.js:42]: No problem!

File: .travis.yml (hunk 7 above)
Comment (tnam-nguyen) [reply .travis.yml:26]: No problem!

File: rest-api/test_backend.py (hunk 8 above)
Comment (tnam-nguyen) [reply rest-api/test_backend.py:73]: I realized that our restAPI does not handle erroneous data for POST requests. Therefore, I updated the POST implementation in sample_backend.py to handle such data. Whenever the user attempts to POST a player with an empty name or a negative score, POST will return 412 Error Status Code with message "Invalid data". I also updated test_backend.py to test for these behaviors. You can check these changes under my last commit "Updated sample_backend.py and test_backend.py to handle invalid data for POST requests."
